NitroxClient/ClientAutoFacRegistrar.cs

```diff
@@ -30,21 +32,23 @@ private static void RegisterCoreDependencies(ContainerBuilder containerBuilder)
                 .SingleInstance();
 
             containerBuilder.RegisterType<PlayerPreferenceManager>().SingleInstance();
-				
+
             containerBuilder.RegisterType<MultiplayerSessionManager>()
                 .As<IMultiplayerSession>()
                 .As<IPacketSender>()
                 .InstancePerLifetimeScope();
 
-            containerBuilder.RegisterType<UdpClient>()
-                .As<IClient>()
+            containerBuilder.RegisterType<LidgrenClient>()
```

Comment:
You can register an implementation as a base class:

```CS
containerBuilder.RegisterType<LidgrenClient>().As<NitroxClient>();
```

Better yet would be to conditionally register one or the other based on a configuration file. You can read more about it here (scroll to lower half of page): https://autofaccn.readthedocs.io/en/latest/configuration/modules.html

Reply:
The problem is that the client falls back to the other client if no connection was established. So both must be registered. Or what do you think exactly?

Reply:
Depends on if you keep the fallback mechanism you have now. If you think it's best like it is then just register them both in autofac. I recommend to still register them as the base class `NitroxClient` though.

Reply:
The general rule of thumb is: 

Register your lowest level abstraction within the IoC container. The reason for that is that the less promises we make in our contracts to the outside world, the less we have to refactor when it comes time to make changes. 

Classes, abstract classes, and Interfaces are all contracts. Interfaces only ever promise signatures in their contracts. Classes and abstract classes; however, make a guarantee of both signatures and behavior (at least for sealed members - which all are by default) in theirs.

In this case, that would be the IClient interface. You'll need to read my review of NitroxClient to understand why that definitely should not be registered in IoC.